pep-0646.rst

```diff
@@ -733,28 +733,166 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
 shape properties of numerical computing programs.
 
 
-Backwards Compatibility
-=======================
+Grammar Changes
+===============
+
+This PEP would require two grammar changes. Full diffs of ``python.gram``
+and simple tests to confirm correct behaviour are available at
+https://github.com/mrahtz/cpython/commits/pep646-grammar.
+
+Star expressions in indexes
+---------------------------
+
+The first grammar change enables use of star expressions in index operations (that is,
+within square brackets), necessary to support star-unpacking of TypeVarTuples.
+
+Before:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.slice+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+After:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.(slice | starred_expression)+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+As with star-unpacking in other contexts, the star operator would call ``__iter__`` on the callee,
+and add the contents of the resulting iterator to the list of index arguments.
+This would enable ``TypeVarTuple`` to be implemented as roughly:
+
+::
 
-In order to use the star operator for unpacking of ``TypeVarTuple`` instances,
-we would need to make two grammar changes:
+    class TypeVarTuple:
+        def __init__(self, name):
+            self._unpacked = _UnpackedTypeVarTuple(name) 
+        def __iter__(self):
+            yield self._unpacked
 
-1. Star expressions must be made valid in at least index operations.
-   For example, ``Tuple[*Ts]`` and ``Tuple[T1, *Ts, T2]`` would both
-   be valid. (This PEP does not allow multiple unpacked ``TypeVarTuple``
-   instances to appear in a single parameter list, so ``Tuple[*Ts1, *Ts2]``
-   would be a runtime error. Also note that star expressions would *not*
-   be valid in slice expressions -  e.g. ``Tuple[*Ts:*Ts]`` is
-   nonsensical and should remain invalid.)
-2. We would need to make '``*args: *Ts``' valid in function definitions.
+There are three things to note about this grammar change. First, for consistency with
+star-unpacking in other contexts, ``_UnpackedTypeVarTuple`` would always be received
+by ``__class_getitem__`` as part of a tuple:
+
+::
+
+    class C:
+        def __class_getitem__(self, x):
+            ...
+            
+    [*(1,)]      # = [1]
+    [*(1,), 2]   # = [1, 2]
+    C[int]       # __class_getitem__ receives x=int
+    # But perhaps surprisingly:
+    C[*Ts]       # __class_getitem__ receives x=(Ts._unpacked, )
+    # This is consistent with:
+    C[*Ts, int]  # __class_getitem__ receives x=(Ts._unpacked, int)
+
+
+Second, the change would also enable star-unpacking of other structures within
+indexing operations:
+
+::
+
+    idxs_to_select = (1, 2, 3)
+    array[0, *idxs_to_select, -1]
```

Comment:
Yes it would, but PEP 637 was rejected because there weren't enough perceived benefits. And honestly the example doesn't make a lot of sense to me. (Is it equivalent to `array[0, 1, 2, 3, -1]`?)

Reply:
I've rewritten this section to make it clearer this is a side effect rather than something we necessarily endorse, and to clarify what this example is equivalent to - better?